### PR TITLE
E2E: Soak test and Functional tests for K8Petstore 

### PR DESCRIPTION
**Pull request by jayunit100**
K8petstore e2e test which runs the k8petstore.sh driver (this PR modifies original driver to play nice with external arguments and so on).

There are two variants.  One is a scalable one (confirms 5K*minions transactions are stored within 6 minutes) and a functional one, which simply confirms that 500 transactions are consumed.  

tested both , results below...

```
TEST PASSED after 7 tries !
Passed...
INFO: Start polling, timeout is 360 seconds
INFO: 2015-05-01 13:43:52.947107496 -0400 EDT ====== 3 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:43:54.949336218 -0400 EDT ====== 3 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:43:56.95111467 -0400 EDT ====== 3 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:43:58.953016636 -0400 EDT ====== 3 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:00.956040158 -0400 EDT ====== 3 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:02.957437582 -0400 EDT ====== 3 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:04.962327399 -0400 EDT ====== 40 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:06.963795521 -0400 EDT ====== 130 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:08.966547325 -0400 EDT ====== 268 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:10.972039863 -0400 EDT ====== 459 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:12.972954366 -0400 EDT ====== 644 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:14.974261224 -0400 EDT ====== 836 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:16.975575057 -0400 EDT ====== 1056 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:18.977949762 -0400 EDT ====== 1318 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:20.979039154 -0400 EDT ====== 1535 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:22.980358649 -0400 EDT ====== 1734 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:24.981367088 -0400 EDT ====== 1889 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:26.982770505 -0400 EDT ====== 1965 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:28.984235613 -0400 EDT ====== 1965 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:30.986099478 -0400 EDT ====== 1965 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:33.045824759 -0400 EDT ====== 2057 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:35.048364096 -0400 EDT ====== 2191 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:37.053733447 -0400 EDT ====== 2425 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:39.055004076 -0400 EDT ====== 2680 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:41.056112696 -0400 EDT ====== 2967 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:43.057228638 -0400 EDT ====== 3235 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:45.058495241 -0400 EDT ====== 3420 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:47.060525703 -0400 EDT ====== 3706 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:49.061816814 -0400 EDT ====== 3946 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:51.063482857 -0400 EDT ====== 4230 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:53.064614873 -0400 EDT ====== 4498 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:55.146951446 -0400 EDT ====== 4716 total petstore transactions stored into redis. =======
INFO: 2015-05-01 13:44:57.154238139 -0400 EDT ====== 4990 total petstore transactions stored into redis. =======
INFO: Done polling, reached timeout 360
[AfterEach] k8bps
  /home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/soak_k8petstore.go:138
STEP: Destroying namespace for this test e2e-tests-k8petstore-soak-4a82c43f-945a-48a6-8bf9-d43fa43d765b

• [SLOW TEST:91.055 seconds]
k8bps
/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/soak_k8petstore.go:170
  k8petstore-SCALE : Should support acquiring up to 1,000,000 transactions
  /home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/soak_k8petstore.go:169
------------------------------

Ran 2 of 43 Specs in 122.626 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 41 Skipped I0501 13:44:59.158509   16912 driver.go:96] All tests pass
```

### CODE REVIEW

**Comment by jayunit100** [comment]
In this patch, ive updated k8petstore and made it amenable to run as an E2E test.

- The e2e test itself runs two tests.  A functional one `k8petstore-functional` and a "soak" test, `k8petstore-SCALE`.  
- The functional test confirms 500 transactions are generated and stored into redis.  This basically is a validation of the connectivity of the containers / persistence ability of redis.  
- The scale test confirms that given 2n data generators, we get 5000 transactions per minion.  So on a ten node cluster, it confirms 50K transactions will be written within a 5 minute time period.    
- This is all done with parameters,  so once its been vetted after a few runs of the CI, I will make the parameters *more demanding*.

**Comment by jayunit100** [comment]
I've rebased this against eric's updates, and also, I've broken commits into logical subunits (the k8petstore examples update is independent from the e2e test).

**Comment by jayunit100** [comment]
not sure why travis failed ?

**Comment by jayunit100** [comment]
Minor note : the soak actually creates 2 load generators per minion... The goal is to confirm that even if each minion is doing generation it should still be able to field rest requests

**Comment by jayunit100** [comment]
bumping my e2e soak test guys @satnam6502 @brendanburns @wojtek-t

**Comment by jayunit100** [comment]
i need to  fix boilerplate and resubmit...

File: test/e2e/soak_k8petstore.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+)
+
+var root0 = absOrDie(filepath.Clean(filepath.Join(path.Base(os.Args[0]), "..")))
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]:
```go
var(
...
)
```

File: test/e2e/soak_k8petstore.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import (
+	"bytes"
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"io/ioutil"
+	"time"
+	//"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: remove dead code.

File: test/e2e/soak_k8petstore.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: separate internal imports from system imports.

File: test/e2e/soak_k8petstore.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"io/ioutil"
+	"time"
+	//"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	. "github.com/onsi/ginkgo"
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: I think there is a preference for _ vs .

File: test/e2e/soak_k8petstore.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+//The IP Below ~ letter->number cipher for k8petstore (most likely it won't be bound by any other process)
+var ip = "165.201.92.15"
+
+// i.e. after 50 trials, expect 3000 transactions... minimum we settle for is 500.
+var minionCount int
+
+// readTransactions reads # of transactions from the k8petstore web server endpoint.
+// for more details see the source of the k8petstore web server.
+func readTransactions(c *client.Client) int {
+
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: delete newline.

File: test/e2e/soak_k8petstore.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	Expect(err).NotTo(HaveOccurred())
+	total_trans, err := strconv.Atoi(string(body))
+	Expect(err).NotTo(HaveOccurred())
+	return total_trans
+}
+
+// runk8petstore runs the k8petstore application, bound to external ip "ip", and polls it to assert that "min_expected"
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: s/runk8petstore/runK8petstore/  (though I love the word 'runk')

File: test/e2e/soak_k8petstore.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+func readTransactions(c *client.Client) int {
+
+	resp, err := http.Get("http://165.201.92.15:3000/llen")
+	if err != nil {
+		log.Fatal("FAILED %v", err) // handle error
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	Expect(err).NotTo(HaveOccurred())
+	total_trans, err := strconv.Atoi(string(body))
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: totalTrans

File: test/e2e/soak_k8petstore.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	Expect(err).NotTo(HaveOccurred())
+	return total_trans
+}
+
+// runk8petstore runs the k8petstore application, bound to external ip "ip", and polls it to assert that "min_expected"
+// transactions are acquired in a maximum of "max_seconds".
+func runK8petstore(ip string, rest_servers int, load_generators int, c *client.Client, min_expected int, max_seconds int64) {
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: camelCase variables.

File: test/e2e/soak_k8petstore.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+		ip,
+		"50", // allow max of 50 trials before we expect the app to be running.
+		strconv.Itoa(rest_servers),
+		strconv.Itoa(load_generators),
+		"1", "0", ns) // 1= # slave, 0 = don't bother running the embedded test.
+
+	var out bytes.Buffer
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]:
```go
if err := cmd.Run(); err != nil {
  ...
}

File: test/e2e/soak_k8petstore.go (hunk 4 above)
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: ah, nm, I see what you are doing, please don't do this, use the package names below.

File: test/e2e/soak_k8petstore.go [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+	if err != nil {
+		fmt.Printf(out.String())
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: you don't appear to write anything to out before printing this, so this print is not needed?

File: test/e2e/soak_k8petstore.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+		log.Fatal(err)
+	}
+
+	//By now, the original k8petstore app has run and spun up a k8petstore w/ load generators.
+	//Lets poll until we reach min_expected transactions
+	total_transactions := 0
+	Logf("Start polling, timeout is %v seconds", max_seconds)
+	for start := time.Now(); time.Since(start).Seconds() < float64(max_seconds); time.Sleep(2 * time.Second) {
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: nit: I prefer:
```go
timeout := time.After(maxSeconds * time.Second)
tick := time.Tick(2 * time.Second)
for {
  select {
    case <-timeout:
      break
    case <-tick:
      ...
  }
}
```

File: test/e2e/soak_k8petstore.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Building a namespace api object")
+		namespace, err = createTestingNS("k8petstore-soak", c)
+		ns = namespace.Name
+		Expect(err).NotTo(HaveOccurred())
+
+		//Now get the # of minions and calibrate the test params to them...
+		minions, err := c.Nodes().List(labels.Everything(), fields.Everything())
+		expectNoError(err)
```
Comment (brendandburns) [inline test/e2e/soak_k8petstore.go]: Expect(err).NotTo(HaveOccurred()) ?

File: test/e2e/soak_k8petstore.go (hunk 3 above)
Comment (jayunit100) [inline test/e2e/soak_k8petstore.go]: gofmt is forcing this on me :(

Comment (jayunit100) [inline test/e2e/soak_k8petstore.go]: Ah ! nevermind.  separation w/ space  solves it

**Comment by jayunit100** [comment]
i fixed most of these, shortly I'm going to make a couple of other improvements that came to mind and resubmit later on.  thanks brendan for the feedback

**Comment by jayunit100** [comment]
I will test this some more, as I made a couple of other enhancements based on some things related to the suggestions (i.e. consolidating variables/ dead code).  I'll Ping when its 100% ready for review.

**Comment by jayunit100** [comment]
okay, tested, ready for a new look

**Comment by brendandburns** [comment]
LGTM.  Please rebase and squash and I'll merge.

Thanks!

**Comment by jayunit100** [comment]
Cool  will rebase shortly ! just saw .

**Comment by jayunit100** [comment]
Squashed down (separate commit for the k8petstore shell script improvement - as its an independent feature enhancement)

**Comment by quinton-hoole** [comment]
This is consistently timing out in our continuous integration environment.  I'm going to revert until that's fixed, possibly as part of #7989

**Pull request merged**